serial: uart_native_pty: ASYNC API support

JordanYates opened the pull request:
Add support for the ASYNC API to the native PTY uart driver.
Asynchronous transmission is performed from the system workqueue.
Asynchronous reception is performed from a dedicated thread.

I would love to add testing for testing this to `tests/drivers/uart/uart_async_api`, but it depends on loopback support.
As far as I can tell the basic polling API is not tested either due to the lack of this support. Does anyone have ideas for enabling this in CI?

JordanYates committed:
serial: uart_native_pty: ASYNC TX support

Add support for transmitting using the asynchronous API. The
asynchronous portion is simulated through the system workqueue.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
serial: uart_native_pty_bottom: length parameter

Add a length parameter to the poll in read function so that data can be
read in larger chunks.

Signed-off-by: Jordan Yates <jordan@embeint.com>

aescolar reviewed (changes requested):
Unfortunately this is not an approach we can use, due to several issues, of which the main are:
1. A signal is delivered fully asynchronously to the native_simulator execution
2. The signal handler will by default be run by *any* native simulator thread
That means between other things that the signal handler will be run, for ex., while the CPU is sleeping corrupting the kernel state; for AMP targets, from any other CPU thread, causing mayhem in both CPUs kernels state, and other messy situations.

In general in native_sim we do not have really asynchronous behaviors. What we do is that we have a deterministic synchronous one that pretends to inject an asynchronous like event at a very deterministic time (see for example the TTY UART interrupt driven mechanism).

dramaticlama commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +255,151 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+
+		event.type = UART_RX_RDY;
+		event.data.rx.buf = buf;
+		event.data.rx.len =  len;
```
Should event.data.rx.len be set to len (bytes asked for) or read (bytes actually read) ?

JordanYates committed:
serial: uart_native_pty: ASYNC RX support

Add support for transmitting using the asynchronous API. The
asynchronous portion is simulated through a dedicated polling thread.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: uart: async_api: added

Add a sample that utilises the ASYNC API to queue packets in bursts.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
> Unfortunately this is not an approach we can use, due to several issues, of which the main are:
> 
>     1. A signal is delivered fully asynchronously to the native_simulator execution
> 
>     2. The signal handler will by default be run by _any_ native simulator thread
>        That means between other things that the signal handler will be run, for ex., while the CPU is sleeping corrupting the kernel state; for AMP targets, from any other CPU thread, causing mayhem in both CPUs kernels state, and other messy situations.
> 
> 
> In general in native_sim we do not have really asynchronous behaviors. What we do is that we have a deterministic synchronous one that pretends to inject an asynchronous like event at a very deterministic time (see for example the TTY UART interrupt driven mechanism).

Thanks for the feedback, I have reworked the PR in line with the TTY driver.
I also added an ASYNC API sample that can be used with the new driver.

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+				data->async.user_callback(data->async.dev, &evt, data->async.user_data);
+			}
+		}
+		if ((data->async.rx_len == 0) || (rc < 0)) {
+			/* Sleep if RX disabled or last read didn't result in any data */
```
why sleep if rx_len == 0? will this code not just end the thread anyhow after that sleep? (and isn't rx_disable() waiting for it to end, so you'd want that to happen ASAP?)
Meaning wouldn't you want to do instead
```suggestion
		if ((data->async.rx_len != 0) && (rc < 0)) {
			/* Sleep if RX is not disabled and last read didn't result in any data */
```

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -48,15 +48,51 @@ struct native_pty_status {
 static int np_uart_init(const struct device *dev);
 
+#ifdef CONFIG_UART_ASYNC_API
+void native_pty_uart_async_poll_function(void *arg1, void *arg2, void *arg3);
```
nit: this proto is not needed?

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+/*
+ * Emulate async interrupts using a polling thread
+ */
+void native_pty_uart_async_poll_function(void *arg1, void *arg2, void *arg3)
```
```suggestion
static void native_pty_uart_async_poll_function(void *arg1, void *arg2, void *arg3)
```
?

aescolar commented on samples/drivers/uart/async_api/src/main.c:
```diff
@@ -0,0 +1,135 @@
+                        /* Allocate the data packet */
+                        tx_buf = net_buf_alloc(&tx_pool, K_FOREVER);
+                        /* Populate it with data */
+                        tx_len = snprintk(tx_buf->data, net_buf_tailroom(tx_buf), "Loop %d: Packet: %d\n", loop_counter, i);
```
```suggestion
                        tx_len = snprintk(tx_buf->data, net_buf_tailroom(tx_buf), "Loop %d: Packet: %d\r\n", loop_counter, i);
```

aescolar replied:
checkpatch is not running(?). max 100 chars per line

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+		CONTAINER_OF(dwork, struct native_pty_status, async.tx_done);
+	struct uart_event evt;
+	unsigned int key = irq_lock();
+	long ret;
+
+	evt.type = UART_TX_DONE;
+	evt.data.tx.buf = data->async.tx_buf;
+	evt.data.tx.len = data->async.tx_len;
+
+	ret = nsi_host_write(data->out_fd, evt.data.tx.buf, evt.data.tx.len);
+	(void)ret;
```
no need for `ret`?
```suggestion

	evt.type = UART_TX_DONE;
	evt.data.tx.buf = data->async.tx_buf;
	evt.data.tx.len = data->async.tx_len;

	(void)nsi_host_write(data->out_fd, evt.data.tx.buf, evt.data.tx.len);
```

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+	struct k_work_sync sync;
+
+	/* Cancel the callback */
+	return k_work_cancel_delayable_sync(&data->async.tx_done, &sync) ? 0 : -EFAULT;
```
I do not know the details of this API, but does the description https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/uart.h#L800
mean that:

* a callback with tx.len = 0 should happen here?

* should this call check for nothing being pended and return -EFAULT  if so?

JordanYates replied:
Callback added, the `k_work_cancel_delayable_sync` returns `false` if the work item has not been submitted, and will therefore return `-EFAULT`.

aescolar commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+
+static int np_uart_tx(const struct device *dev, const uint8_t *buf, size_t len, int32_t timeout)
+{
+	struct native_pty_status *data = dev->data;
```
should this check if the callback is set and return -ENOTSUP otherwise?
(Same for rx and np_uart_tx_abort)

JordanYates replied:
That isn't done in the other drivers that I have inspected

aescolar replied:
Ok, I leave it up to the driver class owner @dcpleung to decide if anything needs doing

aescolar commented on samples/drivers/uart/async_api/src/main.c:
```diff
@@ -0,0 +1,135 @@
+/* change this to any other UART peripheral if desired */
+#define UART_DEVICE_NODE DT_CHOSEN(zephyr_shell_uart)
+
+/* Maximum number of packets to */
```
```suggestion
/* Maximum number of packets to generate per iteration */
```
?

aescolar commented on samples/drivers/uart/async_api/src/main.c:
```diff
@@ -0,0 +1,135 @@
+        case UART_RX_DISABLED:
+                break;
+        case UART_RX_RDY:
+		LOG_HEXDUMP_INF(evt->data.rx.buf + evt->data.rx.offset,
+			evt->data.rx.len, "RX_RDY");
```
silly: Either one tab more or same line?
```suggestion
		LOG_HEXDUMP_INF(evt->data.rx.buf + evt->data.rx.offset,
				evt->data.rx.len, "RX_RDY");
```

aescolar commented on samples/drivers/uart/async_api/src/main.c:
```diff
@@ -0,0 +1,135 @@
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+        struct net_buf *buf;
```
checkpatch is not running(?). tabs for indents only

aescolar reviewed (changes requested):
Thanks @JordanYates 
What I see now is much more minor.
And probably people who is more familiar with the UART API is better than me to provide good input for what remains (CC @dcpleung @nordic-krch )

JordanYates commented:
@aescolar thanks for the feedback, halfway through the implementation I switched from a persistent thread to one that only runs while RX is enabled and missed some things.

nordic-krch commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +247,163 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+	data->async.tx_buf = NULL;
+
+	if (data->async.user_callback) {
+		data->async.user_callback(data->async.dev, &evt, data->async.user_data);
```
Maybe not critical for native but  normally user callback should not be called with interrupts locked.

JordanYates replied:
The goal here was to prevent the system workqueue item being preempted by some other thread while this function was running, since in a "normal" driver this logic would be running from an interrupt context. Happy to use a different solution if you have one.

nordic-krch commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +247,163 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+	}
+
+	/* Generate TX_DONE event with number of bytes transmitted */
+	evt.type = UART_TX_DONE;
```
UART_TX_ABORTED

JordanYates replied:
The UART documentation contradicts itself on which event should be generated:
https://github.com/zephyrproject-rtos/zephyr/blob/6463c68bc394e11beeb7ef0db9afcead9fb3aed2/include/zephyr/drivers/uart.h#L195-L196
https://github.com/zephyrproject-rtos/zephyr/blob/6463c68bc394e11beeb7ef0db9afcead9fb3aed2/include/zephyr/drivers/uart.h#L800

nordic-krch commented on drivers/serial/uart_native_pty.c:
```diff
@@ -209,6 +247,163 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (data->async.rx_len) {
```
the expected receiver behavior is that RX_RDY is reported when buffer is filled or when timeout occurs (and no new data arrives). When buffer is filled and new buffer is not provided then RX_BUF_RELEASED followed by RX_DISABLED events will be generated.

nordic-krch commented on samples/drivers/uart/async_api/src/main.c:
```diff
@@ -0,0 +1,135 @@
+	case UART_RX_BUF_REQUEST:
+		/* Return the next buffer index */
+		LOG_DBG("Providing buffer index %d", async_rx_buffer_idx);
+		rc = uart_rx_buf_rsp(dev, async_rx_buffer[async_rx_buffer_idx],
```
it will assert with current implementation

aescolar replied:
> It won't, because the native driver never issues UART_RX_BUF_REQUEST, since it has no need for multiple buffers.

But this is a generic sample for all UART drivers supporting the ASYNC API, or?

JordanYates replied:
My comment was in reference to what I thought @nordic-krch was referring to, which was that `uart_rx_buf_rsp` would return `-ENOTSUP` for the native driver, and thus trigger the assertion. If that is not the case I would need clarification on the "will assert" comment.

aescolar reviewed (approved):
+2 for the native/host integration side
+1 for driver API/behaviour side
